### ISSUE TITLE
Handle "disappearing messages"

### ISSUE THREAD

**Issue by thisisparker** [thisisparker/tuttle#5]
Currently this app will silently ignore any of Signal's "disappearing message" requests. That should change, though it could happen in a few ways.

For it to comply with disappearing messages, it would have to have more of a sense of `signal-cli`'s contacts model than it currently does. `signal-cli`'s disappearing message support is not super well documented, but I believe information about the setting is set per-contact (not per-message or per-conversation) and so lives in its contacts database.

(Interestingly: supporting disappearing messages could also introduce a mechanism by which messages are, by default, deleted from the inbox after a certain default period that is trivially extendable. Just an idea!)

It could also do some kind of partial support. Like: it could tell you that the other user has requested disappearing messages, but not actually delete them; or it could disable disappearing messages in all conversations it's in; or it could "grey out" messages that are supposed to be disappeared.

**Comment by conorsch** [comment]
>  it could tell you that the other user has requested disappearing messages, but not actually delete them; or it could disable disappearing messages in all conversations it's in

Violating user expectations on disappearing messages sounds like troublesome territories. If a source enables Disappearing Messages, and the app doesn't support that, there should be messaging to explain that behavior (perhaps on the "Landing Page" where this tipline is documented for a given org). 

Forcing support of disappearing messages _for the conversation_ creates a sane default for the source submitting: rather than having the conversation messages/documents lying around on a device indefinitely, the app can force e.g. a 7-day time window, after which submissions will automatically be removed from the submitter's device. Messaging concerns around long-term storage on the org side remain, but that's marginally less risky for the source.

**Comment by thisisparker** [comment]
Yeah, that list was mostly about what is technically possible and includes some verifiably bad ideas. I hadn't really thought all the way through this when I opened the issue, but a default disappearing message setting could interact well with the autoresponder, which could maybe both set the timer and (by default) notify the "source" about the policy, or something. (I think setting the per-contact disappearance setting requires something approximately like sending a message.) I have to see what `signal-cli` supports though, it is especially brittle about that.

**Comment by thisisparker** [comment]
I will say: if this app automatically sets a default disappearing message timer in conversations, that would be a neat nudge in the right direction!

**Comment by conorsch** [comment]
Sounds good! Possibly relevant (also a good contact for hashing out `signal-cli` quirks): https://github.com/guardianproject/haven/issues/249

**Comment by thisisparker** [comment]
Yeah, I remember encountering that :( If I'm reading the [linked issue](https://github.com/AsamK/signal-cli/issues/27) and [corresponding commit](https://github.com/AsamK/signal-cli/commit/82cecfff85ef29e627d4238e3237222b492d5248) and [changelog](https://github.com/AsamK/signal-cli/releases/tag/v0.5.1) correctly, there's not built-in support for _setting_ the timer in `signal-cli` (but it's set in a config file so there is a very janky solution available, of course)

**Comment by thisisparker** [comment]
Yeah, a little more research suggests that assessment is correct: I think `signal-cli` does not actually support setting the disappearing messages timer. (There's [an open issue](https://github.com/AsamK/signal-cli/issues/153) but no progress on it.) So I could still respect disappearing messages, after communicating with `signal-cli`'s contact list, but I couldn't (as of now) set them from the app.

**Comment by thisisparker** [comment]
An update on this: I'm currently deciding between a frequently scheduled job that deletes expired messages OR just deleting them on the first view after they've expired. It's a lot of the same scaffolding, in any case. I'm leaning towards the latter option, which seems a little more robust.

**Comment by thisisparker** [comment]
Not quite closed, but a lot closer. `tuttle` now supports _incoming_ disappearing messages, and deletes them in the app as expected. It does _not_ yet support the deletion of _outgoing_ disappearing messages, although `signal-cli` is appropriately setting the disappearing flag on those messages. Getting support for _outgoing_ disappearing messages is going to require a little more engineering. Anyway, disappearing messages are now indicated with an hourglass icon, and a mouseover shows the expiration information.

![image](https://user-images.githubusercontent.com/400348/50925358-242dc880-1420-11e9-9af3-ec48903fea2f.png)


As is (I believe?) standard in Signal clients, the countdown timer on disappearing messages starts only when the message is first viewed. Less typically, the deletion only happens on pageloads, so if you've got a window open while a message is expiring, it will still be visible until you refresh the page.

**Comment by thisisparker** [comment]
Proper disappearing message support is blocked on an [upstream issue in signal-cli](https://github.com/AsamK/signal-cli/issues/115) ([see also](https://github.com/AsamK/signal-cli/issues/153)). There are a few possible workarounds, but as this tool is primarily for demonstration and prototyping purposes, they will not be implemented here.
